src/labeler.ts

```diff
@@ -6,6 +6,7 @@ import {Minimatch, IMinimatch} from 'minimatch';
 interface MatchConfig {
   all?: string[];
   any?: string[];
+  branch?: string | string[];
```

Comment:
At the moment, the new field (`branch`) is on the same level as the patterns for changed files. The current structure of the configuration file looks like this:
```
- label name:
  - patterns for changed files
  - branch:
    - patterns for branch
```

Potentially, more new fields will be added in the future and such a structure will become non-obvious, non-intuitive. I think we should introduce a new field for changed files as well:
```
- label name:
  - changed files:
    - patterns
  - branch:
    - patterns
... 
```
This will be a breaking change, but we're going to release a new major version anyway to fix the [bug](https://github.com/actions/labeler/pull/480) related to the `sync-labels` input. We can include both changes in the new major release.

Reply:
Interesting so to follow up from your previous comment the we would actually wanna change this to the following:
```yml
- label name:
  - changed-files:
    - patterns
  - head-branch:
    - patterns 
```
So that we could add another level for `base-branch` in the future? 

🤔 Or in the way that `changed-files` can have `any` and `all` we keep that level called `branch` as you suggest here and then add `head` and `base` options to that. So the config might end up looking like this:
```yml
- label name:
  - changed-files:
    - any: 
      - patterns
    - all:
      - patterns
  - branch:
    - head:
      - patterns
    - base:
      - patterns 
```
I like that idea 👍

Reply:
Small follow up question to that - what would we want the defaults to be in the case that these nested options are not provided? 
eg.
```yml
- label 1: 'pattern'
- label 2:
  - changed-files: 'pattern'
- label 3:
  - branch: 'pattern'
```

Currently it defaults to match `any` file that has changed, would we still want to keep that as it is or adjust it? Based on how things are currently I guess it might be something like this:
- label-1 matches any file that has changed?
- label-2 also matches any file that has changed?
- label-3 matches the head branch?

Reply:
In my understanding, the final structure should look like this:
```yml
- label name:
  - changed-files:
    - patterns
    # or
    - any: 
      - patterns
    - all:
      - patterns
  - head-branch:
    - patterns
  - base-branch:
    - patterns 
```
We may want to expand branch settings, for example the base and/or head branches can have `include` or `exclude` options, so I expect them to be separate fields so that we can avoid deep nesting:
```yml
...
  - head-branch:
    - include:
      - patterns
    - exclude:
      - patterns
  - base-branch:
    - include:
      - patterns
    - exclude:
      - patterns
...
```
Regarding your question about omitting nested options - I think we should adhere the following approach:
- `label 1` - throw an error, because it is not obvious what these patterns are related to, what exactly the user meant.
- `label 2` - matches any file that has changed, as you mentioned
- `label 3` - Separating the branches, as I suggested, would simplify the logic:
```yml
  - head-branch:
    - patterns
  - base-branch:
    - patterns 
```

Reply:
Would having include and exclude options more not come from the pattern that you pass? Regardless avoiding deeply nested objects makes sense for now so I'll start on adding that functionality 👍